webapp/src/main/java/com/box/l10n/mojito/service/smartling/ThirdPartyTextUnitMatchingService.java

```diff
@@ -0,0 +1,208 @@
+                }, maxResults
+        );
+        Stream<StringInfo> stream = StreamSupport.stream(sourceStringsPageFetcherSplitIterator, false);
+        List<StringInfo> stringInfoToCheck = stream.collect(Collectors.toList());
+
+
+        logger.debug("done pulling strings for file {}, starting matching", fileUri);
+        Instant start = Instant.now();
+        Set<ThirdPartyTextUnitDTO> thirdPartyTextUnitDTOSet = convertStringInfoToDTO(stringInfoToCheck, fileUri);
+        if (!thirdPartyTextUnitDTOSet.isEmpty()) {
+            List<ThirdPartyTextUnitForBatchImport> thirdPartyTextUnitForBatchImportList = thirdPartyTextUnitSearchService.convertDTOToBatchImport(
+                    thirdPartyTextUnitDTOSet, fileUri.contains(pluralFileIndicator));
+
+            logger.debug("Batch by asset");
+            Map<Asset, List<ThirdPartyTextUnitForBatchImport>> groupedByAsset = thirdPartyTextUnitForBatchImportList
+                    .stream()
+                    .collect(Collectors.groupingBy(ThirdPartyTextUnitForBatchImport::getAsset));
+
+            for (Map.Entry<Asset, List<ThirdPartyTextUnitForBatchImport>> entry : groupedByAsset.entrySet()) {
+                mapThirdPartyTextUnitsToImportWithExisting(entry.getKey(), entry.getValue());
+                importThirdPartyTextUnitOfAsset(entry.getValue());
+            }
+        }
+
+        Instant end = Instant.now();
+        logger.debug("ThirdPartyTextUnit matching and saving start for file {}: {}", fileUri, start);
+        logger.debug("ThirdPartyTextUnit matching and saving end for file {}: {}", fileUri, end);
+    }
+
+    @Transactional
+    void importThirdPartyTextUnitOfAsset(List<ThirdPartyTextUnitForBatchImport> thirdPartyTextUnitsToImport) {
+        logger.debug("Start importing third party text units");
+
+        thirdPartyTextUnitsToImport
+                .stream()
+                .filter(logIfFalse(t -> t.getThirdPartyTextUnitId() != null, logger, "Third party text unit id can't be null, skip: {}", ThirdPartyTextUnitForBatchImport::getThirdPartyTextUnitId))
+                .filter(logIfFalse(t -> t.getTmTextUnit() != null, logger, "Text unit match can't be null, skip: {}", ThirdPartyTextUnitForBatchImport::getThirdPartyTextUnitId))
+                .filter(logIfFalse(this::isUpdateNeededForThirdPartyTextUnit, logger, "Update not needed, skip: {}", ThirdPartyTextUnitForBatchImport::getThirdPartyTextUnitId))
+                .forEach(thirdPartyTextUnitForBatchImport -> {
+                    logger.debug("third party text unit {}: {} corresponding to text unit {} did not find existing match in db",
+                            thirdPartyTextUnitForBatchImport.getThirdPartyTextUnitId(),
+                            thirdPartyTextUnitForBatchImport.getMappingKey(),
+                            thirdPartyTextUnitForBatchImport.getTmTextUnit().getId()
+                    );
+                    ThirdPartyTextUnit thirdPartyTextUnit = new ThirdPartyTextUnit();
```

Comment:
This line can be inside "else", right? ThirdPartyTextUnit object only needs to be created if existingThirdPartyTextUnit not match

Reply:
yes thanks